Implement Server-Side sessions (#68117)
elastic/kibana: docs/user/security/session-management.asciidoc

[role="xpack"]
[[xpack-security-session-management]]
=== Session management

> When you log in to {kib} it creates a session that is used to authenticate subsequent requests to {kib}. A session consists of two components: an encrypted cookie that is stored in your browser, and an encrypted document in a dedicated {es} hidden index. By default, the name of that index is `.kibana_security_session_1` where the prefix is derived from {kib}'s primary `.kibana` index.
> 
> When your session expires, or you log out of {kib} explicitly it will invalidate your cookie and remove session information from the index. {kib} also periodically invalidates and removes any expired sessions that weren't invalidated explicitly.

Reviewer: Suggestion for wording this a bit more consistently:
```suggestion
When you log into {kib}, it creates a session that is used to authenticate subsequent requests. A session consists of two components: an
encrypted cookie that is stored in your browser, and an encrypted document in a dedicated {es} hidden index. By default, the name of that
index is `.kibana_security_session_1`, where the prefix is derived from {kib}'s primary `.kibana` index. If either of these components are
missing, the session is no longer valid.

When you explicitly log out of {kib}, it will delete your session cookie and remove your session document from the index. {kib} also
periodically cleans up the index to remove any expired session documents.
```

[[session-idle-timeout]]
==== Session idle timeout

You can optionally expire sessions after a period of inactivity. By default, a session stays
active until the browser is closed. To define a sliding session expiration, set
the `xpack.security.session.idleTimeout` property in the `kibana.yml`
configuration file. The idle timeout is formatted as a duration of

> `<count>[ms|s|m|h|d|w|M|Y]` (e.g. '70ms', '5s', '3d', '1Y'). For example, set

Reviewer: I'd suggest changing this to match the new description in my top-level comment, if you agree with it.

> You can expire sessions after a period of inactivity. This and `xpack.security.session.lifespan` are both
> highly recommended. By default, session doesn't expire because of inactivity. To define a sliding session expiration, set
> the `xpack.security.session.idleTimeout` property in the `kibana.yml`
> configuration file. The idle timeout is formatted as a duration of
> `<count>[ms|s|m|h|d|w|M|Y]` (e.g. '70ms', '5s', '3d', '1Y'). For example, set
> the idle timeout to expire sessions after 1 hour of inactivity:

Reviewer: It feels odd to see `lifespan` before `idleTimeout`, even though this is the idle timeout section.

How about rewording it like this?

```suggestion
You can use `xpack.security.session.idleTimeout` to expire sessions after a period of inactivity. This and `xpack.security.session.lifespan`
are both highly recommended.

By default, sessions don't expire because of inactivity. To define a sliding session expiration, set the  property in the `kibana.yml`
configuration file. The idle timeout is formatted as a duration of `<count>[ms|s|m|h|d|w|M|Y]` (e.g. '20m', '24h', '7d', '1w'). For example,
set the idle timeout to expire sessions after 1 hour of inactivity:
```
Note 1: changed wording "session doesn't expire" -> "sessions don't expire"
Note 2: changed duration example to match that in prior comments

--
[source,yaml]
--------------------------------------------------------------------------------
xpack.security.session.idleTimeout: "10m"

> --------------------------------------------------------------------------------

Reviewer: How about changing this example to use our planned default of 1 hour?

--

[[session-lifespan]]
==== Session lifespan

You can configure the maximum session duration or "lifespan" -- also known as
the "absolute timeout". By default, a session stays active until the browser is
closed. If an idle timeout is defined, a session can still be extended
indefinitely. To define a maximum session lifespan, set the
`xpack.security.session.lifespan` property in the `kibana.yml` configuration
file. The lifespan is formatted as a duration of `<count>[ms|s|m|h|d|w|M|Y]`

> (e.g. '70ms', '5s', '3d', '1Y'). For example, set the lifespan to expire

Reviewer: I'd suggest changing this to match the new description in my top-level comment, if you agree with it.

> You can configure the maximum session duration or "lifespan" -- also known as
> the "absolute timeout". This and `xpack.security.session.idleTimeout` are both highly
> recommended. By default, session doesn't have a fixed lifespan, and if an idle timeout is defined, a session can still be extended
> indefinitely. To define a maximum session lifespan, set the
> `xpack.security.session.lifespan` property in the `kibana.yml` configuration
> file. The lifespan is formatted as a duration of `<count>[ms|s|m|h|d|w|M|Y]`
> (e.g. '70ms', '5s', '3d', '1Y'). For example, set the lifespan to expire
> sessions after 30 days:

Reviewer: Same as above -- it looked weird to see `idleTimeout` in this section before `lifespan`. How about this?
```suggestion
You can use `xpack.security.session.lifespan` to configure the maximum session duration or "lifespan" -- also known as the "absolute
timeout". This and `xpack.security.session.idleTimeout` are both highly recommended.

By default, sessions don't have a fixed lifespan, and if an idle timeout is defined, a session can still be extended indefinitely. To define
a maximum session lifespan, set the property in the `kibana.yml` configuration file. The lifespan is formatted as a duration of
`<count>[ms|s|m|h|d|w|M|Y]` (e.g. '20m', '24h', '7d', '1w'). For example, set the lifespan to expire sessions after 30 days:
```
Note 1: changed wording "session doesn't have" -> "sessions don't have"
Note 2: changed duration example to match that in prior comments

--
[source,yaml]
--------------------------------------------------------------------------------
xpack.security.session.lifespan: "8h"

> --------------------------------------------------------------------------------

Reviewer: How about changing this example to use our planned default of 30 days?

--

[[session-cleanup-interval]]
==== Session cleanup interval

> You can configure the interval at which {kib} tries to remove expired and invalid sessions from the session index. By default, this value is 1 hour and cannot be less than 10 seconds. To define another interval, set the `xpack.security.session.cleanupInterval` property in the `kibana.yml` configuration file. The interval is formatted as a duration of `<count>[ms|s|m|h|d|w|M|Y]` (e.g. '70000ms', '50s', '3d', '1Y'). For example, schedule session index cleanup to be performed only once a day:

Reviewer 2: ```suggestion
You can configure the interval at which {kib} tries to remove expired and invalid sessions from the session index. By default, this value is 1 hour and cannot be less than 10 seconds. To define another interval, set the `xpack.security.session.cleanupInterval` property in the `kibana.yml` configuration file. The interval is formatted as a duration of `<count>[ms|s|m|h|d|w|M|Y]` (e.g. '70000ms', '50s', '3d', '1Y'). For example, schedule the session index cleanup to perform once a day:
```

============================================================================
If you specify neither session idle timeout nor lifespan, then {kib} will not automatically remove session information from the index unless you explicitly log out. This might lead to an infinitely growing session index. Configure the idle timeout and lifespan settings for the {kib} sessions so that they can be cleaned up even if you don't explicitly log out.
============================================================================

> You can configure the interval at which {kib} tries to remove expired and invalid sessions from the session index. By default, this value is 1 hour and cannot be less than 10 seconds. To define another interval, set the `xpack.security.session.cleanupInterval` property in the `kibana.yml` configuration file. The interval is formatted as a duration of `<count>[ms|s|m|h|d|w|M|Y]` (e.g. '70000ms', '50s', '3d', '1Y'). For example, schedule the session index cleanup to perform once a day:

Reviewer: Another duration example for consistency 🙂 
```suggestion
You can configure the interval at which {kib} tries to remove expired and invalid sessions from the session index. By default, this value is 1 hour and cannot be less than 10 seconds. To define another interval, set the `xpack.security.session.cleanupInterval` property in the `kibana.yml` configuration file. The interval is formatted as a duration of `<count>[ms|s|m|h|d|w|M|Y]` (e.g. '20m', '24h', '7d', '1w'). For example, schedule the session index cleanup to perform once a day:
```

--------------------------------------------------------------------------------
--

[IMPORTANT]
============================================================================

> If you specify neither session idle timeout nor lifespan, then {kib} will not automatically remove session information from the index unless you explicitly log out of {kib}. This may lead to an infinitely growing session index. It's one of the reasons why we strongly recommend configuring idle timeout and lifespan settings for the {kib} sessions so that they can be eventually cleaned up even if you don't log out of {kib} explicitly.

Reviewer 2: ```suggestion
If you don't specify a session idle timeout or lifespan, then {kib} will not automatically remove session information from the index unless you explicitly log out. This might lead to an infinitely growing session index. Configure the idle timeout and lifespan settings for the {kib} sessions so that they can be cleaned up even if you don't explicitly log out.
```

Author: > If you specify neither session idle timeout nor lifespan, then {kib} will not automatically remove session information from the index unless you explicitly log out of {kib}. 
>> If you don't specify a session idle timeout or lifespan, then {kib} will not automatically remove session information from the index unless you explicitly log out. 

@gchaps It seems `If you don't specify a session idle timeout or lifespan` has a different meaning or ..? I wanted to say here that we won't perform cleanup only if _both_ settings are not set, but if any of them is set then it's fine and we'll do the cleanup.

Reviewer 2: Oh you're right.  Ignore my edit then.

> ============================================================================

Reviewer: I feel like this admonition should not be at the bottom of the page.

[[xpack-security-session-management]]
=== Session management

> When you log in to {kib} it creates a session that is used to authenticate subsequent requests to {kib}. A session consists of two components: an encrypted cookie that is stored in your browser, and an encrypted document in a dedicated {es} hidden index. By default, the name of that index is `.kibana_security_session_1` where the prefix is derived from {kib}'s primary `.kibana` index.

Reviewer 2: ```suggestion
When you log in, {kib} creates a session that is used to authenticate subsequent requests to {kib}. A session consists of two components: an encrypted cookie that is stored in your browser, and an encrypted document in a dedicated {es} hidden index. By default, the name of that index is `.kibana_security_session_1`, where the prefix is derived from the primary `.kibana` index.
```

> When your session expires, or you log out of {kib} explicitly it will invalidate your cookie and remove session information from the index. {kib} also periodically invalidates and removes any expired sessions that weren't invalidated explicitly.

Reviewer 2: ```suggestion
When your session expires, or you log out, {kib} will invalidate your cookie and remove session information from the index. {kib} also periodically invalidates and removes any expired sessions that weren't explicitly invalidated.
```

[[session-idle-timeout]]
==== Session idle timeout

You can optionally expire sessions after a period of inactivity. By default, a session stays
active until the browser is closed. To define a sliding session expiration, set
the `xpack.security.session.idleTimeout` property in the `kibana.yml`
configuration file. The idle timeout is formatted as a duration of
`<count>[ms|s|m|h|d|w|M|Y]` (e.g. '70ms', '5s', '3d', '1Y'). For example, set

> the idle timeout to expire idle sessions after 10 minutes of inactivity:

Reviewer 2: ```suggestion
the idle timeout to expire sessions after 10 minutes of inactivity:
```
